Add async Websocket client

mj23000 opened the pull request:
This async version of the Websocket client is mostly just an async version of the existing client, though with the addition of a `RawBaseWebsocketClient` that is inherited from by both clients.

~~No clever handling of the client type has been implemented as with the REST API client with `use_async`. This does have the benefit of not being a breaking change.~~

A similar structure to the existing REST API client has been implemented. 

Commits will be cleaned up when the PR is more ready to be merged.

mj23000 committed:
Add base Websocket class

mj23000 committed:
Add Async Websocket client

GrandMoff100 commented on homeassistant_api/rawwebsocket.py:
```diff
@@ -5,18 +5,17 @@
 
 import websockets.sync.client as ws
 from pydantic import ValidationError
+from rawbasewebsocket import RawBaseWebsocketClient
```
This import needs to be fixed to `from homeassistant_api.rawbasewebsocket ...`. It's causing the docs to fail building. @mj23000

mj23000 committed:
Rework structure and fix typing

mj23000 committed:
Fix typing

GrandMoff100 reviewed (commented):
Looks really awesome! Thank you so much!!!

That just leaves the last thing to do which is to add tests for each of the async websocket methods.

GrandMoff100 commented:
I'm going to convert this to a draft PR, and then you'll convert it back when its ready to be reviewed and merged for real.

mj23000 committed:
Add testing

mj23000 committed:
Use breaks instead of zip(range(1), ...) in async event tests

mj23000 commented:
> Looks really awesome! Thank you so much!!!
> 
> That just leaves the last thing to do which is to add tests for each of the async websocket methods.

That should be it. I can't see the coverage, but hopefully I've covered everything.

GrandMoff100 commented:
> I can't see the coverage

I finally figured out why. Codecov recently changed their behavior so that github actions running on forks now require a `CODECOV_TOKEN` to upload coverage reports (they say it was for ratelimiting purposes, but its annoying). We have a `CODECOV_TOKEN` secret setup to work with that. However, secrets are not passed to forked repo actions by default for security purposes. This repository is configured so that first time contributors require approval to run workflows, and I was under the impression that after approving it, that the workflows would be able to access the secrets for that workflow. But that is not the case, so I just decided to hardcode the token in a config file since this is a public repository, the token is not really sensitive, so it's not the best solution, but its fine for now.

GrandMoff100 commented:
@mj23000  I am happy to aid you in developing the tests we need to increase the coverage. There are some parts of the codebase that need some clever ideas to test them appropriately.

mj23000 commented:
> @mj23000 I am happy to aid you in developing the tests we need to increase the coverage. There are some parts of the codebase that need some clever ideas to test them appropriately.

Yes that sounds good!

adamlogan73 committed:
fixes from rebase

adamlogan73 committed:
Add config entry websocket methods and fix Domain.from_json usage

Use Domain.from_json_with_client() in websocket clients since Domain
requires a client reference. Add config entry methods (get, disable,
enable, ignore flow, subentries, subscribe) to RawWebsocketClient.

adamlogan73 committed:
Add async config entry methods and improve test coverage to 99%

Add async counterparts for all config entry websocket methods in
RawAsyncWebsocketClient. Add async tests for websocket state, entity,
config entry, and error path coverage. Fix conftest return type
annotation, replace zip(range) with break, and use pytest.raises
instead of try/except in error tests.
